python/ray/autoscaler/command_runner.py

```diff
@@ -326,6 +345,163 @@ def _set_ssh_ip_if_required(self):
+                                # messages which SSH puts in stderr
+                                # for no reason
+                                continue
+
+                            if self.timeout_msg_re.\
+                                fullmatch(l) is not None:
+                                # timeout isn't really an error
+                                # but rather a special condition.
+                                #
+                                # it should be handled separately
+                                # and is often expected to occur
+                                if failure_message is not None:
+                                    raise ValueError("Two errors is weird")
+
+                                failure_message="ssh_timeout"
+                                continue
+
+                            if self.conn_refused_msg_re.\
+                                fullmatch(l) is not None:
+                                # this might happen when the instance
+                                # is just starting up, so we want to handle
+                                # it on a case-by-case basis
+                                if failure_message is not None:
+                                    raise ValueError("Two errors is weird")
+
+                                failure_message="ssh_conn_refused"
+                                continue
+
+
+                            if self.known_host_update_msg_re.\
+                                fullmatch(l) is not None:
+                                # this is not an error at all, in fact we
+                                # completely ignore hosts files on purpose
+                                continue
+
+                            if dump_stderr:
+                                output_file.write(l+"\n")
+
+                            cli_logger.error(l)
+
+                return failure_message
+
+            with self.process_runner.Popen(
+                final_cmd,
+                # do not inherit stdin as it messes with signals
+                # (ctrl-C for SIGINT) and these commands aren't
+                # supposed to take input anyway
+                #
+                # P.S. it also messes up the terminal formatting
+                # probably due to some weird stuff that SSH
+                # does to stdin internally if it's a TTY
+                stdin=self.process_runner.PIPE,
+                # fixme: for some reason this is a tty which we DON'T want
+                # commands spam A LOT if they think it's a real tty
+                # especially for progress bars and such
+                # pip is a HUGE issue since it redraws the progress bar
+                # even if no progress was made
+                stdout=self.process_runner.PIPE,
+                stderr=self.process_runner.PIPE,
+                bufsize=1, # line buffering
+                universal_newlines=True # text mode outputs
```

Comment:
OK I guess I still don't know why we have to do this threading work here. What would happen if I just used some simple tee-ing mechanism with a "write" interface?

```
import sys

class Logger(object):
    def __init__(self):
        self.terminal = sys.stdout
        self.log = open("log.dat", "a")

    def write(self, message):
        PROCESS_LINE(message)
        self.terminal.write(message)
        self.log.write(message)  

self.process_runner.Popen(stdout=Logger(), stderr=Logger())
```

Reply:
You can't just pass `Logger` like that because the OS needs a file descriptor to bind to (and Python obtains it by calling `fileno()` on your stream). In general you do need a thread to pump data if you need it to transfer in the background, but it's not clear to me if we need the complexity here.

Reply:
The whole point is reading from both `stdout` and `stderr` simultaneously and separately (we want to tell what came from where). We want this because we want to filter output like `ssh: Connection refused` and `Warning: ... added to hosts file ...` (we deliberately ignore hosts control).

The version that we have in master just throws an exception on non-zero exit code or ignores literally all Exceptions in updater, depending on where the `SSHCommandRunner` was called from. Neither is ideal.

Currently we actually don't really need to snoop `stdout`, but it could be useful in the future. If we decide that the complexity of the threading solution (which turned out to be like 2 lines of code and no locks/other sync primitives) is too much, we can just have a loop for reading stderr and pipe stdout to file as we do now.

Reply:
Also, Richard's example wouldn't work because the python IO that comes from POpen for stdout and stderr is just file descriptor wrappers, so you can only reliably do a blocking read on them. To make things worse, if you don't read from one of these files for too long, the whole process could crash due to internal buffers overflowing.

Reply:
Also `select`-based IO *would work* except there is no support for it on Windows rip

Reply:
@maximsmol I'm still missing high-level context on what's going on unfortunately. (For reference I'm barley familiar with the autoscaler.)
What I'm seeing is a ton of gymnastics with stdin/stdout of an SSH connection to filter out things like progress messages, but I'd expect the solution to that to be to silence the messages at the source (`--quiet `etc.)... so I can't tell why things are the way they are right now. Could you give me some more context as to what's going on without assuming I know much a priori? It's difficult to review this without understanding the reasoning process behind it.